### PR TITLE
[RFC/RDY] make l2tp work with Strongswan 

### PR DESCRIPTION
**Pull request by teto**
###### Motivation for this change
I badly needed l2tp VPN to work for my job so I fought hard with 0 experience on the matter. After a 1-month struggle (https://github.com/NixOS/nixpkgs/issues/30147), here is the PR.

I hope this is how it should be done, I am not sure yet on which nixos files should go where (concerning /etc files).
This is definitely the kind of setup that could use some tests but I've already spent way too much time on this so.

Big thanks to the l2tp maintainer who has been immensely helpful.

###### Things done
- bump l2tp module
- create missing pieces for strongswan to work with l2tp

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/networking/networkmanager.nix [hunk 1]
```diff
@@ -263,7 +263,12 @@ in {
            { source = overrideNameserversScript;
              target = "NetworkManager/dispatcher.d/02overridedns";
            }
-      ++ lib.imap1 (i: s: {
+    ++ optional (cfg.basePackages ? networkmanager_l2tp)
+    {
+      target = "ipsec.d/stub";
+      text = "";
```
Comment (teto) [inline nixos/modules/services/networking/networkmanager.nix]: here I just want to create the ipsec.d folder but didn't know how to do it without creating a stub file.

File: nixos/modules/services/networking/strongswan.nix [hunk 2]
```diff
@@ -114,15 +114,33 @@ in
...
+    # would like to add some path when network manager is enabled
+    secrets2 = cfg.secrets ++ [ "toto" ];
+    secretsFile = ipsecSecrets secrets2;
+  in
+  mkIf enable
+    {
+
+    # with nopeerdns, ppp writes into this folder
+    environment.etc."ppp/stub".text = ''
+    '';
```
Comment (teto) [inline nixos/modules/services/networking/strongswan.nix]: same problem, I dunno how to create just the folder

File: nixos/modules/services/networking/strongswan.nix [hunk 3]
```diff
@@ -114,15 +114,33 @@ in
     };
   };
 
-  config = with cfg; mkIf enable {
+
+  config = with cfg;
+  let
+    # would like to add some path when network manager is enabled
+    secrets2 = cfg.secrets ++ [ "toto" ];
```
Comment (teto) [inline nixos/modules/services/networking/strongswan.nix]: here in principle it should add conditionnally lib.optional networkmanager_l2tp "/etc/ipsec.d/*.secrets" but whatever I do it never makes it into secret2. For instance here 'toto' is not written to the file. I suspect cfg.secrets to partially support the ++ operator (like it discards its content). The PR should not be merged as long as this is not fixed.

**Comment by teto** [comment]
I fixed my previous comments by letting systemd create the folders and believe it's (finally) ready.

**Comment by Mic92** [comment]
Thanks!

**Comment by teto** [comment]
thanks for merging this. I must say when I marked the PR as READY, it did work, I tried it recently and it didn't work anymore and I haven't had the chance to inquire it. The changes you merged are definitely necessary to make it work though. I just note it for other people who might have problems. Maybe the problems were due to other nix components and have been fixed by now. I will check again soon.

**Pull request merged**